cl/phase1/forkchoice/on_block.go

```diff
@@ -89,24 +90,37 @@ func (f *ForkChoiceStore) OnBlock(ctx context.Context, block *cltypes.SignedBeac
 		}
 	}
 
-	var invalidBlock bool
+	//var invalidBlock bool
 	startEngine := time.Now()
 	if newPayload && f.engine != nil {
 		if block.Version() >= clparams.DenebVersion {
 			if err := verifyKzgCommitmentsAgainstTransactions(f.beaconCfg, block.Block.Body.ExecutionPayload, block.Block.Body.BlobKzgCommitments); err != nil {
 				return fmt.Errorf("OnBlock: failed to process kzg commitments: %v", err)
 			}
 		}
-
-		if invalidBlock, err = f.engine.NewPayload(ctx, block.Block.Body.ExecutionPayload, &block.Block.ParentRoot, versionedHashes); err != nil {
-			if invalidBlock {
-				f.forkGraph.MarkHeaderAsInvalid(blockRoot)
+		payloadStatus, err := f.engine.NewPayload(ctx, block.Block.Body.ExecutionPayload, &block.Block.ParentRoot, versionedHashes)
+		switch payloadStatus {
+		case execution_client.PayloadStatusNotValidated:
+			// optimistic block candidate
+			if err := f.optimisticStore.AddOptimisticCandidate(block.Block); err != nil {
+				return fmt.Errorf("failed to add block to optimistic store: %v", err)
 			}
-			return fmt.Errorf("newPayload failed: %v", err)
-		}
-		if invalidBlock {
+			// not sure if need to retry NewPayload for this block later
```

Comment:
because the status is not_validated (don't know if it's valid or not) so I might need to keep this block somewhere to check payload again later? or our entire flow will automatically check it later?

Reply:
i guess dont check it. if you can have chain 1->2->3->4 and 1,2,3,4 are not validated but then we receive 5. then 1->2->3-4->5 are all valid